widgets/src/input/text_selectable.rs

```diff
@@ -1,121 +1,228 @@
 use super::caret_state::CaretPosition;
 use super::CaretState;
-use super::{glyphs_helper::GlyphsHelper, selected_text::SelectedText};
-use crate::layout::{Stack, StackFit};
-use crate::prelude::Text;
+
+use super::glyphs_helper::TextGlyphsHelper;
+use crate::input::glyphs_helper::GlyphsHelper;
+use crate::input::selected_text::SelectedHighLight;
+use crate::prelude::*;
 use ribir_core::prelude::*;
+use std::ops::Range;
 
 #[derive(Declare, Default)]
 pub struct TextSelectable {
   #[declare(default)]
   pub caret: CaretState,
-  #[declare(skip, default)]
-  pub(crate) helper: GlyphsHelper,
+
+  #[declare(skip)]
+  text: CowArc<str>,
+}
+
+pub trait SelectableText {
+  fn selected_text(&self) -> Substr {
+    let rg = self.select_range();
+    self.text().substr(rg)
+  }
+
+  fn select_range(&self) -> Range<usize>;
+
+  fn text(&self) -> &CowArc<str>;
+
+  fn caret(&self) -> CaretState;
+
+  fn set_caret(&mut self, caret: CaretState);
+
+  fn select_text_rect(&self, text: &Text, text_size: Size) -> Vec<Rect> {
```

Comment:
should not pass a `Text` widget, maybe you can provide information by methods like `SelectableText::text`, for example: `TextSelectable::text_style`.